Visual studio suggests removing redundant parenthesis for spans that aren't redundant.

**Version Used**: VS 2019 preview 16.7

**Steps to Reproduce**:
Add the following method to a class.
```cs
private static void BadSuggestion()
{
    var span = (stackalloc byte[8]);
    for(int i = 0; i < span.Length; i++)
    {
         span[i] = (byte)0xFF;
    }
}
```
Visual studio recommends removing the redundant parenthesis in the line `var span = (stackalloc byte[8]);`. However doing so leads to the following errors

> error CS0214: Pointers and fixed size buffers may only be used in an unsafe context
> error CS1061: 'byte*' does not contain a definition for 'Length' and no accessible extension method 'Length' accepting a first argument of type 'byte*' could be found (are you missing a using directive or an assembly reference?)
> error CS0214: Pointers and fixed size buffers may only be used in an unsafe context

**Expected Behavior**:
VS would not suggest to remove the parenthesis.

As an aside, perhaps the issue is that 
```cs
static unsafe void Foo()
{
     var thisIsABytePointer = stackalloc byte[8];
     var thisIsASpanOfBytes = (stackalloc byte[8]);
}
```

--- a/src/Analyzers/CSharp/Tests/RemoveUnnecessaryParentheses/RemoveUnnecessaryExpressionParenthesesTests.cs
+++ b/src/Analyzers/CSharp/Tests/RemoveUnnecessaryParentheses/RemoveUnnecessaryExpressionParenthesesTests.cs
@@ -82,6 +82,20 @@ namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.RemoveUnnecessaryParent
         }
 
         [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryParentheses)]
+        [WorkItem(44629, "https://github.com/dotnet/roslyn/issues/44629")]
+        public async Task TestStackAlloc()
+        {
+            await TestMissingAsync(
+@"class C
+{
+    void M()
+    {
+        var span = (stackalloc byte[8]);
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryParentheses)]
         public async Task TestArithmeticRequiredForClarity2()
         {
             await TestInRegularAndScript1Async(

--- a/src/Analyzers/CSharp/Tests/RemoveUnnecessaryParentheses/RemoveUnnecessaryExpressionParenthesesTests.cs
+++ b/src/Analyzers/CSharp/Tests/RemoveUnnecessaryParentheses/RemoveUnnecessaryExpressionParenthesesTests.cs
@@ -90,7 +90,7 @@ namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.RemoveUnnecessaryParent
 {
     void M()
     {
-        var span = (stackalloc byte[8]);
+        var span = $$(stackalloc byte[8]);
     }
 }");
         }

--- a/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ParenthesizedExpressionSyntaxExtensions.cs
+++ b/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ParenthesizedExpressionSyntaxExtensions.cs
@@ -32,7 +32,6 @@ namespace Microsoft.CodeAnalysis.CSharp.Extensions
                 // https://github.com/dotnet/roslyn/issues/44629
                 return false;
             }
-            
 
             // The 'direct' expression that contains this parenthesized node.  Note: in the case
             // of code like: ```x is (y)``` there is an intermediary 'no-syntax' 'ConstantPattern'

--- a/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ParenthesizedExpressionSyntaxExtensions.cs
+++ b/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ParenthesizedExpressionSyntaxExtensions.cs
@@ -51,7 +51,7 @@ namespace Microsoft.CodeAnalysis.CSharp.Extensions
 
             // Simplest cases:
             //   ((x)) -> (x)
-            if (expression.IsKind(SyntaxKind.ParenthesizedExpression) &&
+            if (expression.IsKind(SyntaxKind.ParenthesizedExpression) ||
                 parentExpression.IsKind(SyntaxKind.ParenthesizedExpression))
             {
                 return true;

--- a/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ParenthesizedExpressionSyntaxExtensions.cs
+++ b/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ParenthesizedExpressionSyntaxExtensions.cs
@@ -61,7 +61,9 @@ namespace Microsoft.CodeAnalysis.CSharp.Extensions
             {
                 // var span = (stackalloc byte[8]);
                 // https://github.com/dotnet/roslyn/issues/44629
-                // PR: https://github.com/dotnet/roslyn/pull/44661
+                // The code semantics changes if the parenthesis removed.
+                // With parenthesis:    variable span is of type `Span<byte>`.
+                // Without parenthesis: variable span is of type `byte*` which can only be used in unsafe context.
                 return false;
             }
 

--- a/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ParenthesizedExpressionSyntaxExtensions.cs
+++ b/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ParenthesizedExpressionSyntaxExtensions.cs
@@ -26,12 +26,6 @@ namespace Microsoft.CodeAnalysis.CSharp.Extensions
             }
 
             var expression = node.Expression;
-            if (expression is StackAllocArrayCreationExpressionSyntax)
-            {
-                // var span = (stackalloc byte[8]);
-                // https://github.com/dotnet/roslyn/issues/44629
-                return false;
-            }
 
             // The 'direct' expression that contains this parenthesized node.  Note: in the case
             // of code like: ```x is (y)``` there is an intermediary 'no-syntax' 'ConstantPattern'
@@ -61,6 +55,14 @@ namespace Microsoft.CodeAnalysis.CSharp.Extensions
                 parentExpression.IsKind(SyntaxKind.ParenthesizedExpression))
             {
                 return true;
+            }
+
+            if (expression is StackAllocArrayCreationExpressionSyntax)
+            {
+                // var span = (stackalloc byte[8]);
+                // https://github.com/dotnet/roslyn/issues/44629
+                // PR: https://github.com/dotnet/roslyn/pull/44661
+                return false;
             }
 
             // (throw ...) -> throw ...

--- a/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ParenthesizedExpressionSyntaxExtensions.cs
+++ b/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ParenthesizedExpressionSyntaxExtensions.cs
@@ -57,7 +57,8 @@ namespace Microsoft.CodeAnalysis.CSharp.Extensions
                 return true;
             }
 
-            if (expression is StackAllocArrayCreationExpressionSyntax)
+            if (expression is StackAllocArrayCreationExpressionSyntax ||
+                expression is ImplicitStackAllocArrayCreationExpressionSyntax)
             {
                 // var span = (stackalloc byte[8]);
                 // https://github.com/dotnet/roslyn/issues/44629

--- a/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ParenthesizedExpressionSyntaxExtensions.cs
+++ b/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ParenthesizedExpressionSyntaxExtensions.cs
@@ -51,7 +51,7 @@ namespace Microsoft.CodeAnalysis.CSharp.Extensions
 
             // Simplest cases:
             //   ((x)) -> (x)
-            if (expression.IsKind(SyntaxKind.ParenthesizedExpression) ||
+            if (expression.IsKind(SyntaxKind.ParenthesizedExpression) &&
                 parentExpression.IsKind(SyntaxKind.ParenthesizedExpression))
             {
                 return true;

--- a/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ParenthesizedExpressionSyntaxExtensions.cs
+++ b/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ParenthesizedExpressionSyntaxExtensions.cs
@@ -26,6 +26,13 @@ namespace Microsoft.CodeAnalysis.CSharp.Extensions
             }
 
             var expression = node.Expression;
+            if (expression is StackAllocArrayCreationExpressionSyntax)
+            {
+                // var span = (stackalloc byte[8]);
+                // https://github.com/dotnet/roslyn/issues/44629
+                return false;
+            }
+            
 
             // The 'direct' expression that contains this parenthesized node.  Note: in the case
             // of code like: ```x is (y)``` there is an intermediary 'no-syntax' 'ConstantPattern'

